### PR TITLE
Totem Timing RecHit Producer

### PR DESCRIPTION
**Pull request by nminafra**
This RP introduces recHits for TotemTiming detector in vertical Roman Pots (90m dedicated run).

- The existing object CTPPSDiamondRecHit was generalized to [CTPPSTimingRecHit](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/DataFormats/CTPPSReco/interface/CTPPSTimingRecHit.h) from which [CTPPSDiamondRecHit](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/DataFormats/CTPPSReco/interface/CTPPSDiamondRecHit.h) and [TotemTimingRecHit](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/DataFormats/CTPPSReco/interface/TotemTimingRecHit.h) inherit.
- Geometry of the new detector added to VeryForwardGeometry: everything new was added in a dedicated directory [TotemTiming](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/Geometry/VeryForwardData/data/TotemTiming/), small changes in the existing code.
- RecHitProducer uses CFD to compute time of arrival from SAMPIC samples in the digis

NOTES:
- the geometry will be added to the DB as soon as it is approved. In the meanwhile it is important to point the reconstruction to the xml geometry. A working example is available [here](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/RecoCTPPS/TotemRPLocal/test/totemTiming_reco_cfg.py).
- With previous hardware configuration, only 2/3 of the detector is connected. The rest will be connected during TS1.
- Suggested run with good statistics: 314276.

NEXT IN LINE:
- fine tuning of the CFD parameters: the parameters in this PR were chosen doing a scan, however a small tuning could be required in the future.
- check cabling (on going).
- DQM (almost ready).
- add SAMPIC calibration: [TotemTimingConversions](https://github.com/CTPPS/cmssw/tree/totemTiming_rechitProducer_PR/RecoCTPPS/TotemRPLocal/interface/TotemTimingConversions.h) was designed to be extensible to read 
calibration values without changing the rest of the code. We aspect a small difference on the performance, tests still on going.

Many thanks for your review!
@forthommel

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

@nminafra 
please confirm that it is possible to still correctly read the CTPPSDiamondRecHit from old files with the new code.

**Comment by nminafra** [comment]
@slava77 I've tested from raw and everything is fine: the interface didn't change.
Then I've run the DQM in a release with the package changed as in this PR, starting from AOD (with "old" rechits) and everything is fine.
Do you propose any other test?
Thanks

**Comment by slava77** [comment]
On 5/1/18 5:45 AM, Nicola Minafra wrote:
> @slava77 <https://github.com/slava77> I've tested from raw and
> everything is fine: the interface didn't change.


> Then I've run the DQM in a release with the package changed as in this
> PR, starting from AOD (with "old" rechits) and everything is fine.

This DQM test sounds good for the purpose.
Just to be sure, was the DQM job  running only an analyzer?
AOD contains digis and I can imagine that reco modules could've just
remade the hits.

> Do you propose any other test?
> Thanks

**Comment by nminafra** [comment]
@slava77 Yes:
process.path = cms.Path(
    process.ctppsDQM
)

For future reference, this is the test I used:
[diamond_dqm_test_cfg.py](https://github.com/CTPPS/cmssw/blob/78c6bfe5d8a2727a92e23d7e890400cf2020db32/DQM/CTPPS/test/diamond_dqm_test_cfg.py)

**Comment by nminafra** [comment]
Forgot to update the CTPPS event content, the test with the error ran fine in my local copy

**Comment by perrotta** [comment]
please test

**Comment by forthommel** [comment]
@perrotta @slava77 
Sorry for the last two extra jenkins builds. We forgot to disable the PPS/Totem local reconstruction from the `FastSim` recipe (again, no sim-level digis to be found in the "classical" `...SIM` event content). This was solved in https://github.com/cms-sw/cmssw/pull/23101/commits/c8c801f7dc8793e240a414fde872fff2d357328f.

**Comment by slava77** [comment]
@cmsbuild please test

@nminafra @forthommel 
thank you for quickly resolving the issue with the fastsim

**Comment by davidlange6** [comment]
Your PR is unmergeable. Please have a look and possibly rebase it.

**Comment by perrotta** [comment]
please test

**Comment by forthommel** [comment]
hold
Trying to figure out what went wrong with this last rebase.

**Comment by forthommel** [comment]
unhold
(fwiw, not sure if the PR issuer has such rights :) )

**Comment by perrotta** [comment]
please test

**Comment by ianna** [comment]
+1

**Comment by perrotta** [comment]
@nminafra @forthommel : could you please point me a test script with an explicit recipe to test, with e.g, run 314276?

File: EventFilter/CTPPSRawToDigi/interface/TotemSampicFrame.h [hunk 1]
```diff
@@ -144,33 +144,31 @@ class TotemSampicFrame
...
+        tmp = *( ( const uint16_t* ) ( totemSampicInfoPtr_ + TotemSampicConstant::timestampA_Position ) );
       }
+      tmp = 0xFFF - tmp;
       return grayToBinary<uint16_t> ( tmp );
     }
 
     inline uint16_t getTimestampB() const {
       uint16_t tmp = 0;
       if ( status_ ) {
-        for ( unsigned short i = 0; i < TotemSampicConstant::timestampB_Size; ++i )
```
Comment (perrotta) [inline EventFilter/CTPPSRawToDigi/interface/TotemSampicFrame.h]: If I am not wrong, all the *_size constants enumerated in TotemSampicConstant now are not used any more. Perhaps you can simplify that ```enum``` by removing all them.

**Comment by nminafra** [comment]
@perrotta The test is available here: [RecoCTPPS/TotemRPLocal/test/totemTiming_reco_cfg.py](https://github.com/CTPPS/cmssw/blob/totemTiming_rechitProducer_PR/RecoCTPPS/TotemRPLocal/test/totemTiming_reco_cfg.py)
(I realize that maybe it was too hidden behind the link "here" in the description of the PR, sorry)

File: RecoCTPPS/TotemRPLocal/src/TotemTimingConversions.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  return firstCellTimeInstant;
+}
+
+//----------------------------------------------------------------------------------------------------
+
+const float TotemTimingConversions::getTriggerTime(const TotemTimingDigi& digi) const
+{
+  unsigned int offsetOfSamples = digi.getEventInfo().getOffsetOfSamples();
+  if (offsetOfSamples == 0)
+    offsetOfSamples = 30; // FW 0 is not sending this, FW > 0 yes
```
[the remark this answers is not in the document]
Comment (nminafra) [reply RecoCTPPS/TotemRPLocal/src/TotemTimingConversions.cc]: This is a parameter in the firmware, it was not sent in the first version of the firmware, but then we decided to send it to add flexibility (indeed we plan to change it). Therefore, if not present (==0) is the value hardcoded in version 0 of the firmware (30). Note that "0" is not a meaningful value for the offset. I anyway took the chance to remove explicit numbers.

File: RecoCTPPS/TotemRPLocal/src/TotemTimingRecHitProducerAlgorithm.cc [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+}
+
+float TotemTimingRecHitProducerAlgorithm::constantFractionDiscriminator(
+    const std::vector<float> &time, const std::vector<float> &data) {
+  std::vector<float> dataProcessed(data);
+  if (lowPassFrequency_ != 0) {
+    // Smoothing
+    for (int i = 0; i < (int)data.size(); ++i) {
+      for (int j = -smoothingPoints_ / 2;
+           j <= +smoothingPoints_; ++j) {
```
[the remark this answers is not in the document]
Comment (nminafra) [reply RecoCTPPS/TotemRPLocal/src/TotemTimingRecHitProducerAlgorithm.cc]: Well spotted! many thanks!

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
Ciao Nicola.

The test crashes with

```
----- Begin Fatal Exception 10-May-2018 11:42:48 CEST-----------------------
An exception of category 'FileInPathError' occurred while
    [0] Constructing the EventProcessor
    [1] Constructing ESSource: class=XMLIdealGeometryESSource  
label='XMLIdealGeometryESSource_CTPPS'
Exception Message:
edm::FileInPath unable to find file  
Geometry/VeryForwardData/data/TotemTiming/TotemTiming_Sens.xml  
anywhere in the search path.
The search path is defined by: CMSSW_SEARCH_PATH
${CMSSW_SEARCH_PATH} is:  
/tmp/aperrott/CMSSW_10_2_X_2018-05-09-2300-testPR23101/src:/tmp/aperrott/CMSSW_10_2_X_2018-05-09-2300-testPR23101/external/slc6_amd64_gcc630/data:/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc6
30/cms/cmssw/CMSSW_10_2_X_2018-05-09-2300/src:/cvmfs/cms-ib.cern.ch/week1/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_X_2018-05-09-2300/external/slc6_amd64_gcc630/data
Current directory is:  
/tmp/aperrott/CMSSW_10_2_X_2018-05-09-2300-testPR23101/src/RecoCTPPS/TotemRPLocal/test
----- End Fatal Exception -------------------------------------------------
```

Please have a look.

In case you need to modify anything in the PR, please consider also  
fixing the indentation in
https://github.com/CTPPS/cmssw/blob/b5f2af6d4577be08bbd0ef1476973a2c412466c7/EventFilter/CTPPSRawToDigi/interface/TotemSampicFrame.h#L22


Nicola Minafra <notifications@github.com> ha scritto:

**Comment by perrotta** [comment]
Sorry for the spam: the comment https://github.com/cms-sw/cmssw/pull/23101#issuecomment-388047046 refers to the version of the PR before your last updates.

As far as I can see, you already noticed and fixed it yourself

**Comment by perrotta** [comment]
+1
- totemTimingLocalReconstruction added to recoCTPPSdets: this adds ```TotemTimingRecHitedmDetSetVector_totemTimingRecHits_*``` to the event. In the test run this adds just 39 (Bytes/Event) to the FEVT/RECO/AOD output
- The previous CTPPSDiamondRecHit  is still there and with the same interface: it now was derives from a base CTPPSTimingRecHit class, from which also the new TotemTimingRecHit derives

**Comment by ianna** [comment]
+1

**Comment by fabiocos** [comment]
@nminafra do I understand correctly that this code is supposed to be valid for the whole 2018 data taking? Clearly, as it is modifying the event content, for its backport in 101X we will need to be careful to manage it though an era, so as not to interfere with the already existing setup. But I would like to understand the situation for 102X

**Comment by nminafra** [comment]
@fabiocos Unless required by you, we do not plan to remove this code. If the detector is not in use, the collections will not be created anyways, so it creates no harm to us.
Actually, in this regard: I was planning to extend also [CTPPSDiamondLocalTracks](https://github.com/cms-sw/cmssw/blob/master/DataFormats/CTPPSReco/interface/CTPPSDiamondLocalTrack.h) in a similar way (mother class CTPPSTimingLocalTrack, inheriting CTPPSDiamondLocalTrack and TotemTimingLocalTrack).  I didn't do it because I'm not sure when the track producer will be ready. However, if you want, I can do it now in a way that, for the backport, both changements can be done in the same era.
Thanks

**Comment by slava77** [comment]
On 5/11/18 1:37 AM, Nicola Minafra wrote:
> @fabiocos <https://github.com/fabiocos> Unless required by you, we do
> not plan to remove this code. If the detector is not in use, the
> collections will not be created anyways, so it creates no harm to us.
> Actually, in this regard: I was planning to extend also
> CTPPSDiamondLocalTracks
> <https://github.com/cms-sw/cmssw/blob/master/DataFormats/CTPPSReco/interface/CTPPSDiamondLocalTrack.h>
> in a similar way (mother class CTPPSTimingLocalTrack, inheriting
> CTPPSDiamondLocalTrack and TotemTimingLocalTrack). I didn't do it
> because I'm not sure when the track producer will be ready. However, if
> you want, I can do it now in a way that, for the backport, both
> changements can be done in the same era.
> Thanks
> 

There is going to be a rereco with 10_2_X.
All CTPPS rereco can still be redone by running on AOD.
So, we shouldn't push too hard on 10_1_X backport beyond the operational
needs for CTPPS to take good quality data during the upcoming special run.

BTW, regarding "If the detector is not in use, the collections will not
be created ": all configured reco producer modules should produce a
collection every event. It can be empty.

**Comment by nminafra** [comment]
> BTW, regarding "If the detector is not in use, the collections will not
be created ": all configured reco producer modules should produce a
collection every event. It can be empty.

True! thanks for the clarification @slava77

**Comment by nminafra** [comment]
Thanks everyone for the review!

**Comment by slava77** [comment]
On 5/14/18 11:23 AM, Fabio Cossutti wrote:
> @slava77 <https://github.com/slava77> I am thinking to the strategy to
> be adopted for the 90 m beta* data taking of end of June, when the full
> 10_2_X will not still fully at production quality.
> 
> —

What I meant is:
If the hits are needed just for analysis, their reconstruction can be
done on the fly from AOD files at analysis level.
So, we do not have to change 10_1_X production setup.
That's unless the rechits are required to take data (e.g. in the trigger)

**Comment by nminafra** [comment]
@slava77 @fabiocos this PR (and the connected ones) is needed for the dedicated run for both DQM and prompt reco: we don't need the rechits in any other (past) run.

**Pull request merged**